### DISCUSSION TITLE
Segmentation of one dataset with 4 different CNN models (UNet, DynUNet, SegResNet, and VNet)

### DISCUSSION THREAD

**Discussion by Kiarashdnsh** [Project-MONAI/MONAI#6707] [Q&A] [answered]
Hello,

I'm a novice in AI and image segmentation.
I have a dataset of [256, 256, 256]-sized images and labels. The input channel is 1 and the output channel is 3 (label IDs are 0, 1, and 2, respectively). I'm trying to segment the images using UNet, DynUNet, VNet, and SegResNet with the MONAI framework.
For UNet, I have used the following code:

```python
model = UNet(
        spatial_dims=3,
        in_channels=1,
        out_channels=3,
        channels=(16, 32, 64, 128, 256),
        strides=(2, 2, 2, 2),
        num_res_units=2,
        norm=Norm.BATCH,
    ).to(device)
loss_function = DiceLoss(include_background=False, to_onehot_y=True, softmax=True)
optimizer = torch.optim.Adam(model.parameters(), 1e-4)
``` 
and used this tutorial [here ](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/torch/unet_training_dict.py) for running the trainer.

Thus far, the results have been satisfactory on the UNet front! However, I also wanted to use the same dataset with the same dataloader, but with a different network model in this framework, starting with DynUNet. To achieve this, I changed the model to:

```python
model = DynUNet(
        spatial_dims=3,
        in_channels=1,
        out_channels=3,
        kernel_size=(3, 3, 3),
        strides=(2, 2, 2),
        upsample_kernel_size=(2, 2, 2),
        deep_supervision=False,
        filters=(2, 2, 2),
        dropout=0.2,
        norm_name="instance",
        act_name="leakyrelu",
        deep_supr_num=1,
        res_block=True,
        trans_bias=True,
)
lloss_function = DiceLoss(include_background=False, to_onehot_y=True, sigmoid=True)
optimizer = torch.optim.Adam(model.parameters(), 1e-4)
``` 

But now I get this error while running the same trainer [code](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/torch/unet_training_dict.py):

```python
AssertionError: ground truth has different shape (torch.Size([2, 2, 256, 256, 256])) from input (torch.Size([2, 2, 128, 128, 128]))
``` 

- Can you help me solve this issue?
- What can I do to issuelessly implement other CNN models in the same environment?

Thanks in advance,
Kiarash

**Comment by KumoLiu** [comment, marked as answer]
Hi @Kiarashdnsh, thanks for your interest here.
- In your case, you downsample the image twice and upsample only once, so your output shape is different from the label. You can simply update the stride to (1, 2, 2).
And for more details about how to set the parameters of the `DynUNet`, you can refer to 
https://github.com/Project-MONAI/MONAI/blob/17c1e3acff583e241066da08dae4b75554b52dda/monai/networks/nets/dynunet.py#L70-L85

- You can replace the networks in this pipeline with any network seamlessly, but you need to at least understand the structure of each network and know how to set the parameters to get the desired output size.

Hope it can help you, thanks!

**Reply by Kiarashdnsh** [reply]
Thanks. With your help, I managed to solve the issue. But where do I find more info on the structure of different models and the functionality of their hyper/parameters? The Monai's documentation is not detailed enough for someone at my level. For example, I don't know the what the `sw_batch_size` is in the training pipeline, and how it is connected with the `roi_size` parameter, and the documentation doesn't provide in-depth info on it. 
I need information on many of these parameters and their functionality. Thanks! 🙌

**Reply by KumoLiu** [reply]
Hi @Kiarashdnsh, you can print the network architecture or maybe just read the original paper we cited. I think most of the models in MONAI have citations like these.
https://github.com/Project-MONAI/MONAI/blob/17c1e3acff583e241066da08dae4b75554b52dda/monai/networks/nets/dynunet.py#L58-L61
Thanks!
